src/mesh/amr_loadbalance.cpp

```diff
@@ -462,368 +697,183 @@ void Mesh::RedistributeAndRefineMeshBlocks(ParameterInput *pin, ApplicationInput
   int nbs = nslist[Globals::my_rank];
   int nbe = nbs + nblist[Globals::my_rank] - 1;
 
-#ifdef MPI_PARALLEL
-  int bnx1 = GetBlockSize().nx1;
-  int bnx2 = GetBlockSize().nx2;
-  int bnx3 = GetBlockSize().nx3;
-  // Step 3. count the number of the blocks to be sent / received
-  Kokkos::Profiling::pushRegion("Step 3: Count blocks");
-  int nsend = 0, nrecv = 0;
-  for (int n = nbs; n <= nbe; n++) {
-    int on = newtoold[n];
-    if (loclist[on].level > newloc[n].level) { // f2c
-      for (int k = 0; k < nleaf; k++) {
-        if (ranklist[on + k] != Globals::my_rank) nrecv++;
+  // Step 7 - eps: Restrict fine to coarse buffers
+  for (int on = onbs; on <= onbe; on++) {
+    int nn = oldtonew[on];
+    if (newloc[nn].level < loclist[on].level) {
+      const IndexDomain interior = IndexDomain::interior;
+      auto pmb = FindMeshBlock(on);
+      IndexRange cib = pmb->c_cellbounds.GetBoundsI(interior);
+      IndexRange cjb = pmb->c_cellbounds.GetBoundsJ(interior);
+      IndexRange ckb = pmb->c_cellbounds.GetBoundsK(interior);
+      // Need to restrict this block before doing sends
+      for (auto &var : pmb->vars_cc_) {
+        if (var->IsAllocated()) {
+          pmb->pmr->RestrictCellCenteredValues(var.get(), cib.s, cib.e, cjb.s, cjb.e,
+                                               ckb.s, ckb.e);
+        }
       }
-    } else {
-      if (ranklist[on] != Globals::my_rank) nrecv++;
     }
   }
+  Kokkos::fence();
+
+#ifdef MPI_PARALLEL
+  // Step 7. Send data from old to new blocks
```

Comment:
We got two Steps 7 now

Reply:
We have 7 - eps and 7 now ;). 

But we also no longer have steps 3 through 6, I think I will just remove the Step comments.